crates/containerd-shim-wasm/src/sandbox/cgroups.rs

```diff
@@ -0,0 +1,674 @@
+    pub fn new(base: PathBuf, path: PathBuf) -> Self {
+        CgroupV2 {
+            base: base,
+            path: path,
+        }
+    }
+
+    fn get_file(&self, name: &str) -> PathBuf {
+        self.base.join(&self.path).join(name)
+    }
+}
+
+fn ensure_write_file(path: std::path::PathBuf, content: &str) -> Result<()> {
+    let parent = path.parent().unwrap();
+    if !parent.exists() {
+        create_dir_all(parent).map_err(|e| {
+            Error::Others(format!(
+                "could not create parent cgroup dir {}: {}",
+                parent.to_str().unwrap(),
+                e
+            ))
+        })?;
+    }
+    let mut file = OpenOptions::new().write(true).open(&path).map_err(|e| {
+        Error::Others(format!(
+            "could not open cgroup file {}: {}",
+            path.to_str().unwrap(),
+            e
+        ))
+    })?;
+    file.write_all(content.as_bytes())?;
+    Ok(())
+}
+
+impl Cgroup for CgroupV2 {
+    fn version(&self) -> Version {
+        Version::V2
+    }
+
+    fn open(&self) -> Result<RawFD> {
+        let path = self.base.join(&self.path);
+        create_dir_all(&path)?;
+        nix::fcntl::open(
+            path.as_path(),
+            nix::fcntl::OFlag::O_DIRECTORY
+                | nix::fcntl::OFlag::O_CLOEXEC
+                | nix::fcntl::OFlag::O_RDONLY,
+            nix::sys::stat::Mode::empty(),
+        )
+        .map_err(|e| std::io::Error::from(e).into())
+    }
+
+    fn delete(&self) -> Result<()> {
+        let path = self.base.join(&self.path);
+        if path.exists() {
+            std::fs::remove_dir(path)?;
+        }
+        Ok(())
+    }
+
+    fn add_task(&self, pid: u32) -> Result<()> {
+        ensure_write_file(self.get_file("cgroup.procs"), &format!("{}", pid))?;
+        Ok(())
+    }
+
+    // See https://github.com/containerd/cgroups/blob/724eb82fe759f3b3b9c5f07d22d2fab93467dc56/v2/utils.go#L164
+    // for details on converting the oci spec (which is heavily v1 focussed) to v2.
+    // Also https://github.com/containers/crun/blob/2497b9bb03623838d37a1587087f1ad3d6ff28ec/crun.1.md#cgroup-v2
+    fn apply(&self, res: Option<Resources>) -> Result<()> {
+        if res.is_none() {
+            return Ok(());
+        }
+        let res = res.unwrap();
+
+        if let Some(cpu) = res.cpu() {
+            if let Some(shares) = cpu.shares() {
+                let s = 1 + ((shares - 2) * 9999) / 262142;
+                ensure_write_file(self.get_file("cpu.weight"), &format!("{}", s))?;
+            }
+            let mut max = "max".to_string();
+            if let Some(quota) = cpu.quota() {
+                max = format!("{}", quota);
+            }
+            if let Some(period) = cpu.period() {
+                ensure_write_file(self.get_file("cpu.max"), &format!("{} {}", max, period))?;
+            } else {
+                ensure_write_file(self.get_file("cpu.max"), max.as_str())?;
+            }
+
+            // no realtime support
```

Comment:
It's just a note since the oci spec has "real-time" cpu constraints but cgroupv2 does not support that (though we do include in in the cgroupv1 code).
I can delete it or possibly expand on this.